### PR TITLE
Rewrite grid synchronization

### PR DESCRIPTION
**Pull request by csarofeen**
Modified grid synchronization to enable multi grid reduction code that was previously erroring, reverting restrictions we had that https://github.com/csarofeen/pytorch/pull/1215 addressed. The new grid semaphore is reusable in a kernel with cooperative launches, I've validated it, but will need more changes to enable cooperative kernels.

### CODE REVIEW

File: torch/csrc/jit/codegen/cuda/runtime/block_sync_default.cu [hunk 1]
```diff
@@ -10,3 +10,57 @@ __forceinline__ __device__ void sync() {
...
+__device__ int64_t check_global(volatile int64_t& semaphore) {
+  return semaphore;
+}
+
+template <bool X_BLOCK, bool Y_BLOCK, bool Z_BLOCK>
+__device__ void grid_sync(
+    int64_t& semaphore,
+    const bool& persistent,
+    const uint64_t& segment_size,
+    bool& last_block) {
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/runtime/block_sync_default.cu]: Can you please add comments about these parameters?

File: torch/csrc/jit/codegen/cuda/lower_thread_predicate.cpp [hunk 2]
```diff
@@ -392,7 +330,7 @@ ParallelTypeBitmap ThreadPredicateMap::getParallelBroadcastDomains(
     if (!id->isBroadcast()) {
       continue;
     }
-    if (id->isBlockDim() || (!output_smem && id->isThreadDim())) {
+    if (id->isThread() || (!output_smem && id->isThreadDim())) {
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/lower_thread_predicate.cpp]: Why is this changed? If a tensor is on shared memory, it doesn't need any specific broadcast operation other than just storing and syncthreads.

**Comment by naoyam** [comment]
Looks good except for the one change in `lower_thread_predicate.cpp`.

**Comment by naoyam** [comment]
I get a validation error in `FusionZeroDimBroadcast`:

```
[ RUN      ] NVFuserTest.FusionZeroDimBroadcast_CUDA
unknown file: Failure
C++ exception with description "aten_output_tensor.allclose( fusion_output_tensor.to(aten_output_tensor.dtype()), tolerance_values.second, tolerance_values.first)INTERNAL ASSERT FAILED at "../test/cpp/jit/test_gpu_validator.h":400, please report a bug to PyTorch.

Validation error in output 0 on line 7058 in file ../test/cpp/jit/test_gpu.cpp.
  Detected abs error of: 3.3915e-05
    absolute tolerance was set to 1.30594e-05
    and relative tolerance set to 1.30594e-07
Exception raised from testValidate at ../test/cpp/jit/test_gpu_validator.h:400 (most recent call first):
```

The error is pretty small, so it seems just a rounding error. The test shouldn't be affected by any of the changes in this PR. I don't see the error when the test is run in isolation.

File: torch/csrc/jit/codegen/cuda/lower_thread_predicate.cpp (hunk 2 above)
Comment (csarofeen) [reply torch/csrc/jit/codegen/cuda/lower_thread_predicate.cpp]: Thank you, good catch.

**Comment by csarofeen** [comment]
I changed the test to be a parallel reduction to pass the tolerance more consistently (wasn't failing for me)

File: torch/csrc/jit/codegen/cuda/type.h [hunk 3]
```diff
@@ -194,12 +194,12 @@ static constexpr std::array<ParallelType, 6> kParallelTypeThreads = {
...
+static constexpr std::array<ParallelType, 3> kParallelTypeBIDs = {
     ParallelType::BIDx,
     ParallelType::BIDy,
     ParallelType::BIDz};
 
-static constexpr std::array<ParallelType, 6> kParallelTypeTIDs = {
+static constexpr std::array<ParallelType, 3> kParallelTypeTIDs = {
     ParallelType::BIDx,
     ParallelType::BIDy,
     ParallelType::BIDz};
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/type.h]: Can you also change `BID` to `TID`? Looks like this static variable isn't used yet, so nothing isn't impacted (fortunately).

**Review by naoyam — APPROVED** [review]
LGTM. Commented on a small bug that's not part of this PR.
